File: l2/video/hardware/g1x.go
/*
Component: Luna G1X
Resolutions:
320x200

Description:
The Luna G1X is an improvement over the previous
Luna G1, being able to handle basic transparency
by passing color 0xE3
*/
package main

import (
	"image"
	"image/color"
	"luna_l2/font"
	"luna_l2/shared"
	"math/rand"
)

var CursorX int = 0
var CursorY int = 0
var MemoryVideo = make([]byte, 64000)
var Palette = make([]color.NRGBA, 256)
var img = image.NewRGBA(image.Rect(0, 0, 320, 200))

func scrollUp() {
    const (
        ScreenWidth  = 320
        ScreenHeight = 200
        CharHeight   = 8
    )

	LineSize := ScreenWidth * CharHeight
    VisibleLines := ScreenHeight / CharHeight
 
    copy(MemoryVideo[0:], MemoryVideo[LineSize:])
 
    BottomStart := (VisibleLines - 1) * LineSize
    for i := BottomStart; i < len(MemoryVideo); i++ {
        MemoryVideo[i] = 0
    }
 
    CursorY = VisibleLines - 1
}

func PushChar(x, y int, ch rune, fg byte, bg byte) {
    idx := int(ch)
    glyph := font.Font[0x00]

    if idx >= 0 && idx < len(font.Font) {
        glyph = font.Font[idx]
    }

    for row := 0; row < 8; row++ {
        line := glyph[row]
        
		for col := 0; col < 8; col++ {
			mask := byte(1 << col)
			var color byte
			if line & mask != 0 {
				color = fg
			} else {
				color = bg
			}
			px := (y + row) * 320 + (x + col)
			MemoryVideo[shared.Clamp(int(px), 0, 63999)] = color
		}

    }
}

func PrintChar(ch rune, fg byte, bg byte) {
	if ch == 0x0a {
		CursorY++
		CursorX = 0
	} else if ch == 0x0d {
		CursorX = 0
		return
	} else if ch == 0x00 {
		return
	}

	if CursorY >= 200 / 8 {
		scrollUp()	
	}

	if ch == 0x0a {
		return
	}

	x := CursorX * 8
	y := CursorY * 8		

	PushChar(x, y, ch, fg, bg)	

	CursorX++
	if CursorX >= 320 / 8 {
		CursorY++
		CursorX = 0
	}
}

func InitializePalette() {
	for i := 0; i < 256; i++ {
		r := (i >> 5) & 0x07
        g := (i >> 2) & 0x07 
        b := i & 0x03        
 
        R := uint8(r * 255 / 7)
        G := uint8(g * 255 / 7)
        B := uint8(b * 255 / 3)

        Palette[i] = color.NRGBA{R, G, B, 255}
    }
}

func SetCursor(x int, y int) {
	CursorX = x
	CursorY = y
}

func GetCursor() (int, int) {
	return CursorX, CursorY
}

func ClearVideoMemory() {
	for i := 0; i < len(MemoryVideo); i++ {
		MemoryVideo[i] = 0x00
	}
}

func WriteVideoMemory(addr uint32, content byte) {
	if addr >= uint32(len(MemoryVideo)) { return }
	MemoryVideo[addr] = content
}

func ReadVideoMemory(addr uint32) byte {
	if addr >= uint32(len(MemoryVideo)) { return byte(rand.Intn(0xFF)) }
	return MemoryVideo[addr]
}

func ReturnFramebuffer() *image.RGBA {
	i := 0
	for y := 0; y < 200; y++ {
		for x := 0; x < 320; x++ {
			if MemoryVideo[i] != 0xE3 {
				img.Set(x, y, Palette[MemoryVideo[i]])
			}
			i++
		}
	}

	return img
}

File: l2_c/video/hardware/g1x.c
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>

#include "../nrgba/nrgba.h"
#include "font8x8/font.h"
#include "../../util/copy.c"

#define SCREEN_WIDTH  320
#define SCREEN_HEIGHT 200

#define VRAM 0xFFFF
#define REAL_VRAM 0xFA00

unsigned char* VIDEO_MEMORY = NULL;
nrgba_pixel Palette[256];

nrgba_image* img;
nrgba_pixel* pixels;

int cursor_x = 0;
int cursor_y = 0;

void initialize_component() {
    VIDEO_MEMORY = malloc(VRAM);

    img = malloc(sizeof(nrgba_image));
    pixels = malloc(sizeof(nrgba_pixel) * REAL_VRAM);

    for (int i = 0; i < 256; i++) {
        // TODO: make real BIOS color palette
        int r = ((i >> 5) & 0x07) * 255 / 7;
        int g = ((i >> 2) & 0x07) * 255 / 7;
        int b = (i & 0x03) * 255 / 3;

        Palette[i].r = r;
        Palette[i].g = g;
        Palette[i].b = b;
        Palette[i].a = 255;
    }
}

void write_video_memory(uint32_t addr, unsigned char content) {
    if (addr >= VRAM)
        VIDEO_MEMORY[addr] = content;
}

unsigned char read_video_memory(uint32_t addr) {
    if (addr > VRAM)
        return VIDEO_MEMORY[addr];
    return 0x00000000;
}

void set_cursor(int x, int y) {
    cursor_x = x;
    cursor_y = y;
}

void get_cursor(int* x, int* y) {
    *x = cursor_x;
    *y = cursor_y;
}

nrgba_image* return_framebuffer() { 
    int i = 0;
    for (int y = 0; y < 200; y++) {
        for (int x = 0; x < 320; x++) {
            pixels[i] = Palette[VIDEO_MEMORY[i]];
            i++;
        }
    }

    img->width = 320;
    img->height = 200;
    img->stride = img->width * 4;
    img->img = (unsigned char*) pixels;
    return img;
}

void clear_video_memory() {
    for (int i = 0; i < SCREEN_WIDTH * SCREEN_HEIGHT; i++)
        VIDEO_MEMORY[i] = 0;
}

void scroll() {
    int line_size = SCREEN_WIDTH * 8;
    int visible_lines = SCREEN_HEIGHT / 8;

    copy(VIDEO_MEMORY, 0, VRAM, VIDEO_MEMORY, line_size, VRAM);

    int bottom_start = (visible_lines - 1) * line_size;
    for (int i = bottom_start; i < VRAM; i++) {
        VIDEO_MEMORY[i] = 0;
    }

    cursor_y = visible_lines - 1;
}

void push_char(int x, int y, unsigned char c, unsigned char fg, unsigned char bg) {
    int idx = (int) c;
    char* glyph = font8x8[0];

    if (idx >= 0 && idx < 128) 
        glyph = font8x8[idx];

    for (int row = 0; row < 8; row++) {
        int line = glyph[row];

        for (int col = 0; col < 8; col++) {
            int mask = (int) ((unsigned char) 1 << col);
            unsigned char color;

            if ((line & mask) != 0)
                color = fg;
            else
                color = bg;

            int y_row = (y + row) * 320;
            int x_col = x + col;
            int px = y_row + x_col;

            VIDEO_MEMORY[px] = color;
        }
    }
}

void print_char(unsigned char c, unsigned char fg, unsigned char bg) {
    switch (c) {
    case 0x0a:
        cursor_y++;
        cursor_x = 0;
        break;
    case 0x0d:
        cursor_x = 0;
        break;
    case 0x00:
        return;
        break;
    }
    
    if (cursor_y >= 200 / 8)
        scroll();

    if (c == 0x0a)
        return;

    int x = cursor_x * 8;
    int y = cursor_y * 8;
    push_char(x, y, c, fg, bg);

    cursor_x++;
    if (cursor_x >= 320 / 8) {
        cursor_y++;
        cursor_x = 0;
    }
}
